client/pluginmanager/csimanager/volume_test.go

```diff
@@ -597,3 +598,113 @@ func TestVolumeManager_InterleavedStaging(t *testing.T) {
 	must.Eq(t, 1, csiFake.NodeUnpublishVolumeCallCount, must.Sprint("expected 1 unpublish call"))
 	must.Eq(t, 0, csiFake.NodeUnstageVolumeCallCount, must.Sprint("expected no unstage call"))
 }
+
+func TestVolumeManager_Serialization(t *testing.T) {
+	ci.Parallel(t)
+
+	tmpPath := t.TempDir()
+	csiFake := &csifake.Client{}
+
+	logger := testlog.HCLogger(t)
+
+	ctx := hclog.WithContext(t.Context(), logger)
+
+	manager := newVolumeManager(logger,
+		func(e *structs.NodeEvent) {}, csiFake,
+		tmpPath, tmpPath, true, "i-example")
+
+	// ensure that serialized ops for the same volume cannot interleave.
+
+	ctx, cancel := context.WithTimeout(t.Context(), time.Second)
+	t.Cleanup(cancel)
+	errs := make(chan error, 3)
+	results := make(chan string, 3)
+
+	// we can't guarantee the goroutines will contend, so we'll force the first
+	// op to wait until the other goroutines have at least started, and then
+	// sleep for a bit to give them a chance to catch up
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			wg.Wait()
+			time.Sleep(100 * time.Millisecond)
+			results <- "one"
+			return errors.New("one")
+		})
+	}()
+
+	go func() {
+		wg.Done() // unblock the first op
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			results <- "two"
+			return errors.New("two")
+		})
+	}()
+
+	go func() {
+		wg.Done() // unblock the first op
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			results <- "three"
+			return errors.New("three")
+		})
+	}()
+
+	found := 0
+	for {
+		if found >= 3 {
+			break
+		}
+		select {
+		case err := <-errs:
+			found++
+			select {
+			case result := <-results:
+				if err.Error() != result {
+					t.Fatal("got result out-of-order from error")
+				}
```

Comment:
just to be sure I understand: since each of these channels (`errs` and `results`) have a buffer size of `3`, and there are 3 competing goroutines, this ensures that whatever the buffers end up with, the order of `errs` matches the order of `results`, as would happen if they were forced to run in serial.

the idea being that it would identify any racy gap between `results <- "aaa"` and `return errors.New("aaa")`

let's say the func doesn't do what it's supposed to (I am sure that it does), I guess this would be flaky, because they would sometimes interleave and sometimes not?

Reply:
Except now I'm realizing this is actually flaky 🤦 ... once we've exited the serialized function there's no guarantee that the send on the error channel isn't going to be out-of-order if one of the other goroutines returns quickly enough.

Lemme go back to the drawing board with this.